igorpex/aws-nodejs-be
# Task 9 Backend For Frontend

1.   FE PR\REPO: https://github.com/igorpex/shop-react-redux-cloudfront/pull/7
2.   FE URL: http://igorpex-rs-app.s3-website-eu-west-1.amazonaws.com/
3.   Product Service API URL:  https://a3loi3zsk8.execute-api.eu-west-1.amazonaws.com/dev
4.   Cart Service API URL: http://igorpex-cart-api-develop.eu-west-1.elasticbeanstalk.com/api
5.   BFF Service API URL: http://igorpex-bff-api-develop.eu-west-1.elasticbeanstalk.com
6. Example of how to call createProduct labmda with all needed information: URL, payload, headers, etc.:
This is example with BFF, you can change BFF URL `http://igorpex-bff-api-develop.eu-west-1.elasticbeanstalk.com` to direct Lambda URL  `https://a3loi3zsk8.execute-api.eu-west-1.amazonaws.com/dev`
```
fetch("http://igorpex-bff-api-develop.eu-west-1.elasticbeanstalk.com/products", {
  "headers": {
    "accept": "application/json, text/plain, */*",
    "content-type": "application/json",
  },
  "body": "{\"imageurl\":\"https://picsum.photos/id/237/300/200\",\"price\":700,\"description\":\"New Old Black Dog from Harry Potter book\",\"title\":\"New Old Black Dog\",\"count\":\"1\"}",
  "method": "POST",
});
```
7. Example how to call Product Service and Cart Service via BFF Service URL
7.1. Get one product:
`http://igorpex-bff-api-develop.eu-west-1.elasticbeanstalk.com/products/904e8e59-8ea3-45f3-931d-fd5faadb4b18`
7.2. Get all products
`http://igorpex-bff-api-develop.eu-west-1.elasticbeanstalk.com/products`
7.3. Get cart: 
`http://igorpex-bff-api-develop.eu-west-1.elasticbeanstalk.com/profile/cart`

Done: 07.08.22 / Deadline  07.08.22
8. Score: 6/7
- **3** - A working and correct `Express` application should be in the `bff-service` folder. Reviewer can start this application locally with any valid configuration in the `.env` file and this application should works as described in the Task 9.1
- **5** - The BFF Service should be deployed with Elastic Beanstalk. The BFF Service call should be redirected to the appropriate service : Product Service or Cart Service. The response from the BFF Service should be the same as if Product Service or Cart Service services were called directly.
- **+1**-Caching 2 minutes for GET requests

**To test batch import:**
1. use authorization_token aWdvcnBleDpURVNUX1BBU1NXT1JE in localStorage.
2. Use file 3NewProducts.xls from 'bff-service/test-products' folder. 
3. New Black Dog, New Flying car, New Old Bicycle will be added.

You can manage(edit), delete any product, you can add products one by one.

Keep in mind cache (2 mins), it will not allow you to see change results immidiately.

My discord igorbo#2208 (or @igorpex )

Screenshots:
Create product (request to BFF):
![create-product](https://user-images.githubusercontent.com/7703384/183397180-17c4ea47-803f-4179-bc76-3ad00e69de93.jpg)
Edit product (request to BFF):
![edit-product](https://user-images.githubusercontent.com/7703384/183397208-0de405f5-8662-4c15-a5c6-6a385b15034b.jpg)
Batch Import (request to normal lambda, as /import path out of BFF scope)
![batch-import](https://user-images.githubusercontent.com/7703384/183397281-093ae22c-78f2-4ff3-acf5-d45e6ea1ef62.jpg)
Cart (request to BFF)
![cart](https://user-images.githubusercontent.com/7703384/183397471-19ebca47-5c35-403f-96d9-e4cb12cef4d6.jpg)
